File: backend/api/loadpull.py
"""Load Pull export.

Filed under `/test` so it shares the frontend's existing dev-server proxy, but
kept in its own module: unlike the sweep, this test's loop runs in the browser,
so there is no run state here to export from. The rows come up with the request.
"""
from __future__ import annotations

import asyncio
import logging
import time

from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import Response

from ..loadpull import (
    LoadPullExportRequest, LoadPullRow, build_workbook, parse_workbook,
)
from .errors import handle_driver_errors

log = logging.getLogger(__name__)

router = APIRouter(prefix="/test/load-pull", tags=["load-pull"])

XLSX_MEDIA_TYPE = (
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
)

#: A load pull is positions x frequencies x powers; tens of thousands of rows
#: would already be an unusable spreadsheet.
MAX_ROWS = 100_000


@router.post("/export")
async def export(req: LoadPullExportRequest) -> Response:
    if not req.rows:
        raise HTTPException(status_code=400, detail="No results to export")
    if len(req.rows) > MAX_ROWS:
        raise HTTPException(status_code=413, detail=f"Too many rows (max {MAX_ROWS})")

    with handle_driver_errors("load pull export"):
        # openpyxl rendering is CPU-bound and grows with the row count.
        data = await asyncio.to_thread(build_workbook, req.rows, req.meta)

    filename = time.strftime("load_pull_%Y%m%d_%H%M%S.xlsx")
    return Response(
        content=data,
        media_type=XLSX_MEDIA_TYPE,
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


#: A Load Pull workbook is a few thousand rows at most.
MAX_IMPORT_BYTES = 25 * 1024 * 1024


@router.post("/import", response_model=list[LoadPullRow])
async def import_xlsx(request: Request) -> list[LoadPullRow]:
    """Read a previously exported workbook back into rows.

    Raw body rather than a multipart upload: that needs no `python-multipart`
    dependency, and the client has a single file to send.
    """
    data = await request.body()
    if not data:
        raise HTTPException(status_code=400, detail="No file received")
    if len(data) > MAX_IMPORT_BYTES:
        raise HTTPException(status_code=413, detail="File too large")

    try:
        return await asyncio.to_thread(parse_workbook, data)
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e)) from e
    except Exception as e:
        raise HTTPException(status_code=422, detail=f"Could not read the workbook: {e}") from e

File: frontend/src/api/loadPull.ts
import { http, httpBlob } from './client'
import type { LoadPullResultRow } from '../store/loadPullPageStore'

/** Run parameters for the workbook's Run sheet — none of it is in the rows. */
export interface LoadPullExportMeta {
  freq_spec?: string
  power_spec?: string
  settle_ms?: number
  pa_mode?: number
  delta_x_mm?: number
  zero_pulses?: number | null
  end_pulses?: number | null
  path_loss_default_db?: number
  path_loss_points?: Array<{ freq_mhz: number; db: number; calibrated: boolean }>
  dut_mac?: string | null
}

export const loadPullApi = {
  /**
   * Render the results to a workbook.
   *
   * The rows go up with the request: this test's loop runs in the browser, so
   * the backend has no copy of them — it is being used for openpyxl, not for
   * state it holds.
   */
  exportXlsx: (rows: LoadPullResultRow[], meta: LoadPullExportMeta) =>
    httpBlob('/test/load-pull/export', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ rows, meta }),
    }),

  /** Read an exported workbook back into rows. */
  importXlsx: (file: File) =>
    http<LoadPullResultRow[]>('/test/load-pull/import', {
      method: 'POST',
      body: file,
      headers: { 'Content-Type': 'application/octet-stream' },
    }),
}
